### PR TITLE
feat!: Remove CodePath#currentSegments

### LINKED ISSUE #17457: Change Request: Deprecate and remove `CodePath#currentSegments`
### ESLint version

HEAD

### What problem do you want to solve?

Currently, the `CodePath` objects produced via `onCodePathStart` and `onCodePathEnd` contain a mix of data and state. The most glaring example is `CodePath#currentSegments`, which is actually a live array that is updated as `CodePathAnalyzer` traverses the AST. (`CodePath#currentSegments` actually is a passthrough to an underlying `CodePathState#currentSegments` property.)

This is problematic with respect to the [language plugins `SourceCode#traverse()`](https://github.com/eslint/rfcs/blob/main/designs/2022-languages/README.md#the-sourecodetraverse-method), which allows for a traversal to happen at any point and returns an array of steps to complete the traversal. The data available at each step should be the same as when there's an active traversal, but that doesn't happen with code path analysis due to the `CodePath` objects exposing state information.

So, exposing state information on `CodePath` objects is a blocker for the language plugins proposal.

 

### What do you think is the correct solution?

As best I can tell, the only problematic property is `CodePath#currentSegments`, which I think we should deprecate in  the v8 timeframe and see if we can remove it in either v9 or v10. From a search through the codebase it seems that this is the primary property we are using off of `CodePath` objects, and it doesn't appear that other state information that changes during traversal is being exposed.

Further, I doubt that there are many people using `CodePath#currentSegments` in the wild, so I think we can be pretty aggressive with removing it. I searched through some of the popular plugins and found zero instances of code path usage.

The `currentSegments` property is easily duplicated inside of rules using this code:

```js
const codePathSegments = [];
let currentCodePathSegments = [];

return {
            onCodePathStart() {
                codePathSegments.push(currentCodePathSegments);
                currentCodePathSegments = [];
            },
            onCodePathSegmentStart(segment) {
                currentCodePathSegments.push(segment);
            },
            onCodePathSegmentEnd() {
                currentCodePathSegments.pop();
            },
            onCodePathEnd() {
                currentCodePathSegments = codePathSegments.pop();
            },
};
```

I'd argue that this is the most appropriate way to track current traversal state (from within a rule rather than inside of the core).

### Participation

- [X] I am willing to submit a pull request for this change.

### Additional comments

_No response_

### PR DESCRIPTION
**Pull request by nzakas**
#### Prerequisites checklist

- [ ] I have read the [contributing guidelines](https://github.com/eslint/eslint/blob/HEAD/CONTRIBUTING.md).

#### What is the purpose of this pull request? (put an "X" next to an item)

[ ] Documentation update
[ ] Bug fix ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/bug-report.md))
[ ] New rule ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/rule-proposal.md))
[ ] Changes an existing rule ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/rule-change-proposal.md))
[ ] Add autofix to a rule
[ ] Add a CLI option
[x] Add something to the core
[ ] Other, please explain:

#### What changes did you make? (Give an overview)

Removed `CodePath#currentSegments` along with associated checks in `RuleTester` and `FlatRuleTester`. Also updates related documentation.

fixes #17457

### CODE REVIEW

File: tests/lib/rule-tester/rule-tester.js [hunk 1]
```diff
@@ -2490,31 +2490,117 @@ describe("RuleTester", () => {
...
 
-        it("should emit a deprecation warning when CodePath#currentSegments is accessed", () => {
+        it("should pass-through services from parseForESLint to the rule and log deprecation notice", () => {
+            const enhancedParserPath = require.resolve("../../fixtures/parsers/enhanced-parser");
+            const disallowHiRule = {
+                create: context => ({
+                    Literal(node) {
+                        assert.strictEqual(context.parserServices, context.sourceCode.parserServices);
+
+                        const disallowed = context.sourceCode.parserServices.test.getMessage(); // returns "Hi!"
```
Comment (mdjermanovic) [inline tests/lib/rule-tester/rule-tester.js:2501]:
```suggestion
```

File: tests/lib/rule-tester/rule-tester.js [hunk 2]
```diff
@@ -2490,31 +2490,117 @@ describe("RuleTester", () => {
...
-            const useCurrentSegmentsRule = {
-                create: () => ({
-                    onCodePathStart(codePath) {
-                        codePath.currentSegments.forEach(() => {});
+                        if (node.value === disallowed) {
+                            context.report({ node, message: `Don't use '${disallowed}'` });
+                        }
                     }
                 })
             };
```
Comment (mdjermanovic) [inline tests/lib/rule-tester/rule-tester.js:2508]:
```suggestion
```

File: tests/lib/rule-tester/rule-tester.js [hunk 3]
```diff
@@ -2490,31 +2490,117 @@ describe("RuleTester", () => {
...
+                        parser: enhancedParserPath,
+                        errors: [{ message: "Don't use 'Hi!'" }]
+                    }
+                ]
             });
 
             assert.strictEqual(processStub.callCount, 1, "calls `process.emitWarning()` once");
             assert.deepStrictEqual(
                 processStub.getCall(0).args,
                 [
```
Comment (mdjermanovic) [inline tests/lib/rule-tester/rule-tester.js:2528]:
```suggestion
```

File: tests/lib/rule-tester/rule-tester.js [hunk 4]
```diff
@@ -2490,31 +2490,117 @@ describe("RuleTester", () => {
...
+                        `"deprecated-method" rule is using \`context.${methodName}()\`, which is deprecated and will be removed in ESLint v9. Please use \`sourceCode.${replacementName}()\` instead.`,
+                        "DeprecationWarning"
+                    ]
+                );
+            });
+
         });
 
+
     });
```
Comment (mdjermanovic) [inline tests/lib/rule-tester/rule-tester.js:2604]:
```suggestion
    });
```

**Review by mdjermanovic — COMMENTED** [review]
Seems that there's some extra code from previous versions.

For reference, here's the original commit from https://github.com/eslint/eslint/pull/17756: https://github.com/eslint/eslint/commit/7f0e31fd843cd91feadb89af45303fd097953e03

**Comment by mdjermanovic** [comment]
Can you try applying the remaining ones?

**Comment by mdjermanovic** [comment]
Oh, I got it, the suggestions conflicted with each other.

**Comment by nzakas** [comment]
Yes, sorry, that's what I meant. :) I'll circle back to this one once alpha is squared away.

**Comment by mdjermanovic** [comment]
Here's another attempt: https://github.com/eslint/eslint/pull/17936

**Comment by nzakas** [comment]
Closing in favor of #17936

**Pull request closed without merging**